bin/packages/build-worker.js

```diff
@@ -0,0 +1,159 @@
+const writeFile = promisify( fs.writeFile );
+
+/**
+ * Promisified sass.render.
+ *
+ * @type {Function}
+ */
+const renderSass = promisify( sass.render );
+
+/**
+ * Get the package name for a specified file
+ *
+ * @param  {string} file File name
+ * @return {string}      Package name
+ */
+function getPackageName( file ) {
+	return path.relative( PACKAGES_DIR, file ).split( path.sep )[ 0 ];
+}
+
+/**
+ * Get Build Path for a specified file.
+ *
+ * @param  {string} file        File to build
+ * @param  {string} buildFolder Output folder
+ * @return {string}             Build path
+ */
+function getBuildPath( file, buildFolder ) {
+	const pkgName = getPackageName( file );
+	const pkgSrcPath = path.resolve( PACKAGES_DIR, pkgName, 'src' );
+	const pkgBuildPath = path.resolve( PACKAGES_DIR, pkgName, buildFolder );
+	const relativeToSrcPath = path.relative( pkgSrcPath, file );
+	return path.resolve( pkgBuildPath, relativeToSrcPath );
+}
+
+/**
+ * Object of build tasks per file extension.
+ *
+ * @type {Object<string,Function>}
+ */
+const BUILD_TASK_BY_EXTENSION = {
+	async '.scss'( file ) {
+		const outputFile = getBuildPath( file.replace( '.scss', '.css' ), 'build-style' );
+		const outputFileRTL = getBuildPath( file.replace( '.scss', '-rtl.css' ), 'build-style' );
+
+		const [ , contents ] = await Promise.all( [
+			makeDir( path.dirname( outputFile ) ),
+			readFile( file, 'utf8' ),
+		] );
+
+		const builtSass = await renderSass( {
+			file,
+			includePaths: [ path.resolve( __dirname, '../../assets/stylesheets' ) ],
+			data: (
+				[
+					'colors',
+					'breakpoints',
+					'variables',
+					'mixins',
+					'animations',
+					'z-index',
+				].map( ( imported ) => `@import "${ imported }";` ).join( ' ' )	+
+				contents
+			),
+		} );
+
+		const result = await postcss( require( './post-css-config' ) ).process( builtSass.css, {
+			from: 'src/app.css',
+			to: 'dest/app.css',
+		} );
+
+		const resultRTL = await postcss( [ require( 'rtlcss' )() ] ).process( result.css, {
+			from: 'src/app.css',
+			to: 'dest/app.css',
+		} );
+
+		await Promise.all( [
+			writeFile( outputFile, result.css ),
+			writeFile( outputFileRTL, resultRTL.css ),
+		] );
+	},
+
+	async '.js'( file ) {
+		for ( const [ environment, buildDir ] of Object.entries( JS_ENVIRONMENTS ) ) {
+			const destPath = getBuildPath( file, buildDir );
+			const babelOptions = getBabelConfig( environment, file.replace( PACKAGES_DIR, '@wordpress' ) );
+
+			const [ , transformed ] = await Promise.all( [
+				makeDir( path.dirname( destPath ) ),
+				babel.transformFileAsync( file, babelOptions ),
+			] );
+
+			await Promise.all( [
+				writeFile( destPath + '.map', JSON.stringify( transformed.map ) ),
+				writeFile( destPath, transformed.code + '\n//# sourceMappingURL=' + path.basename( destPath ) + '.map' ),
+			] );
+		}
+	},
+};
+
+module.exports = async ( file, callback ) => {
+	const extension = path.extname( file );
+	const task = BUILD_TASK_BY_EXTENSION[ extension ];
```

Comment:
Noting for future: I've been planning in my head how I might implement something like #13124 . I don't plan to make the changes here, since it's already a fairly significant refactor.

I expect this task could be separated out into a function `buildFile`, which is optionally enhanced to read from / write to a cache. I'm thinking in some cases we may want to opt out of the caching behavior, e.g. in `npm run package-plugin`, since it seems there could be a (very-low-probability-but-possible) chance that hash collisions could yield a bad result†. That said, we should always be testing (manually and automated) the artifact, but as of today, we're only doing the former.

```js
let buildTask = () => { /* ... */ };

if ( hasCache ) {
	buildTask = firstMatch( [ fromCache, buildTask ] );
}

buildTask = flow( [ buildTask, toCache ] );
```

† This assumes that an implementation would be similar to `babel-loader` in checking e.g. a cache folder for a file named by the MD5 (or other fast hash) of the source file's content.

cc @noisysocks

Reply:
Perhaps store the cache in a directory within the repository that is `.gitignore`d? That way, when `npm run package-plugin` runs `git clean -fxd` it will empty the cache. 

Not sure I'd worry about collisions. I'd use SHA1 as it's fast, has a low collision rate, and is good enough for git's object storage model.

Reply:
> Perhaps store the cache in a directory within the repository that is `.gitignore`d? That way, when `npm run package-plugin` runs `git clean -fxd` it will empty the cache.

That's true, I'd not considered that. My plan was to do similar to what's done with `babel-loader`, i.e. a folder at `node_modules/.cache`. Seems it would work just the same as you propose (since `node_modules` is `gitignore`d).

> Not sure I'd worry about collisions. I'd use SHA1 as it's fast, has a low collision rate, and is good enough for git's object storage model which what inspired my approach in #13124.

I'm not overly worried, I'll just inevitably feel uncomfortable about the lack of being able to say that it's "impossible". I don't feel too strongly about the specific hash algorithm, just that it's fast. Some of what I've read seemed to imply MD5 beats out SHA1 on this front 🤷‍♂ [[1]](http://atodorov.org/blog/2013/02/05/performance-test-md5-sha1-sha256-sha512/)[[2]](https://stackoverflow.com/questions/2722943/is-calculating-an-md5-hash-less-cpu-intensive-than-sha-family-functions)[[3]](https://www.cryptopp.com/benchmarks.html)